src/plugins/intel_cpu/src/nodes/gather.cpp

```diff
@@ -649,13 +649,13 @@ void Gather::execCompressed4Bit() {
 
                 OUT_TYPE* pdst = &dstData[dstIdx];
 
-                const size_t scale_offset = srcIdx / scale_group_size;
-                auto cur_zp = have_zp ? zp[srcIdx / zp_group_size] : 0;
                 size_t p = srcIdx;
                 size_t dst_idx = 0;
 
                 for (; p < srcIdx + afterAxisSize; p++) {
                     auto val = srcData[p >> 1];
+                    const size_t scale_offset = p / scale_group_size;
+                    auto cur_zp = have_zp ? zp[p / zp_group_size] : 0;
```

Comment:
Such change will negatively affect the perf given more compute moved into innermost loop.
I would propose to split the loop into 2 nested ones, Smt like:
```
group_size = min(scale_group_size, zp_group_size)
groups_num = (srcIdx + afterAxisSize) / group_size 
for (blocks_num):
    scale = ... 
    zp = ...
    for (group_size):
       pdst[dst_idx] = static_cast<OUT_TYPE>((get4Bit(val, p % 2) - zp) * scale);
```

Reply:
Done，
hi @dmitry-gorokhov 
1: If `axis` != 0, your suggestion will not work. but `axis` always equal to 0 for all cased I saw.
2: For most real case, zp is empty or zp is 1D, or zp and scale have same group size,  or group size equal to 1.
Based on this heuristic, I split them into some branches and process them separately.